---
title: "vignette-Illumina4"
author: "Xutao Wang"
package: curatedTBData
output: BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{"vignette-Illumina4"}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}  
---

```{r setup, include=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
if (! "SummarizedExperiment" %in% installed.packages()) BiocManager::install("SummarizedExperiment")
if (! "MultiAssayExperiment" %in% installed.packages()) BiocManager::install("MultiAssayExperiment")
if (! "GEOquery" %in% installed.packages()) BiocManager::install("GEOquery")
if (! "dplyr" %in% installed.packages()) install.packages("dplyr")
if (! "plyr" %in% installed.packages()) install.packages("plyr")
if (! "TBSignatureProfiler" %in% installed.packages()) devtools::install_github("aubreyodom/TBSignatureProfiler")
```

```{r libraries, message = FALSE}
library(GEOquery)
library(SummarizedExperiment)

#library(limma)
library(affy)
library(dplyr)
# library(plyr)
devtools::load_all(".")
library(curatedTBData)
#library(TBSignatureProfiler)
#library(ggplot2)
#library(gridExtra)
#library(ggridges)
#library(illuminaHumanv3.db) # Illumina HumanHT-12 V3.0 expression beadchip
#library(illuminaHumanv4.db) # Illumina HumanHT-12 V4.0 expression beadchip
```

# Create Summarized Experiment Object from Illumina 4.0
## Read in Non-normalized data from 
```{r obtain data information illumina4, cache = TRUE, results="hide"}
geo_access <- "GSE39939"
dat <- download_data_Illumina4(geo_access)
head(dat)
```

## Create Row Data
```{r create row data illumina4, message=FALSE}
# use function from the package
row_data <- create_RowData(dat,"GPL10558")
head(row_data)[1:5,1:5]
```


## Create Column Data
```{r download sample information illumina4,message = FALSE}
gse <- getGEO(geo_access, GSEMatrix = FALSE)
```

```{r create column data illumina4,results="hide"}
col_data <- create_ColData(geo_access,gse)
colnames(col_data) <- c('TBStatus','HIVStatus','GeographicalRegion')

# Rename TB Status
relabel_TB <- function(dat_new){
  dat_new$culture_status <- sub(".*\\((.*)\\).*", "\\1", dat_new$TBStatus)
  
  dat_new$culture_status <- ifelse(dat_new$culture_status==dat_new$TBStatus,'n.a.',dat_new$culture_status)
  
  TBStatus <- TBStatus_temp <- dat_new$TBStatus
  
  for (i in 1:length(TBStatus)){
  if (TBStatus[i] == unique(TBStatus_temp)[1] || TBStatus[i] == unique(TBStatus_temp)[4]){
    TBStatus[i] = 'PTB'
  }
    if (TBStatus[i] == unique(TBStatus_temp)[2] || TBStatus[i] == unique(TBStatus_temp)[5]){
    TBStatus[i] = 'OD'
  }
    if (TBStatus[i] == unique(TBStatus_temp)[3]){
    TBStatus[i] = 'Latent'
  }
  }
  dat_new$TBStatus <- TBStatus
  return(dat_new)
}

col_data <- relabel_TB(col_data)
head(col_data)
```

### Match Desscription ID to Sample ID
Why this step: Colnames from the imported data is not the sample ID
```{r get corresponding sample ID illumina4}
description_id_raw <- sapply(1:length(names(GSMList(gse))),
                             function(x) GSMList(gse)[[x]]@dataTable@columns$Column[3])
# Demo: Convert 6116725094_J.Detection Pval to 6116725094_J
description_id <- sapply(1:length(description_id_raw), 
                         function(x) gsub("\\..*", "", description_id_raw[x]))
ID_table <- data.frame(SampleID = names(GSMList(gse)), DescriptionID = description_id)
head(ID_table)
```

```{r mapping DescriptionID to sampleID illumina4}
# Demo: convert X6247215025_L.AVG_Signal to 6247215025_L, should use \\. when mathcing . 
colnames(dat) <- sapply(1:ncol(dat), 
                             function(x) gsub(".*X|\\..*", "", colnames(dat)[x]))
indx <- sapply(1:length(ID_table$DescriptionID), 
              function(i) which(colnames(dat) %in% ID_table$DescriptionID[i]))

dat <- dat[,indx]
colnames(dat) <- ID_table$SampleID
head(dat)[1:5,1:5]

```

## Create Summazried Experiment Obeject
```{r create summazried experiment object illumina4}
# Remove files in temp directory
# unlink(paste0(normalizePath(tempdir()), "/", dir(tempdir())), recursive = TRUE)

experimentData <- new('MIAME',
                      name="Victoria Wright",
                      lab="Wright Fleming Institute",
                      contact="v.wright@imperial.ac.uk",
                      title="Diagnosis of childhood tuberculosis and host RNA expression in Africa",
                      abstract="Children were recruited from 2 hospitals in Coast Province, Kenya (n=157) who were either HIV+ or HIV - with either active TB (culture confirmed), active TB (culture negative), LTBI or OD.",
                      url="10.1056/NEJMoa1303657",
                      pubMedIds = '24785206',
                      other=list(Platform = 'Illumina HumanHT-12 V4.0 expression beadchip (GPL10558)'))

GSE39939_sobject <- SummarizedExperiment(assays = list(counts= as.matrix(dat)), colData = col_data, rowData = row_data, metadata = list(experimentData)); GSE39939_sobject
```

# Create Summarized Experiment Object from Illumina 3.0
## Read in Non-normalized data
```{r obtain data information illumina3, cache = TRUE, results="hide"}
geo_access <- "GSE19442"
dat_list <- download_data_Illumina3(geo_access)

# Combine list to dataframe
dat_non_normalized <- create_data_Illumina3(dat_list)
```

## Create Row Data
```{r create row data illumina3,message=FALSE}
# use function from the package
row_data <- create_RowData(dat_non_normalized,"GPL6947")
head(row_data)[1:5,1:5]
```

## Create Column Data
```{r download sample information illumina3,message = FALSE}
gse <- getGEO(geo_access, GSEMatrix = FALSE)
```

```{r}
col_data <- create_ColData(geo_access,gse)
colnames(col_data) <- c('Age','Gender','Ethnicity','TBStatus','GeographicalRegion','BcgVaccinated','BirthRegion','TST', 'SputumSmearStatus','sputum_culture')

indx <- sapply(1:length(names(GSMList(gse))), function(i) 
  which(names(GSMList(gse)) %in% colnames(dat_non_normalized)[i]));indx
col_data <- col_data[indx,]

row.names(col_data) <- colnames(dat_non_normalized)

# Rename TB Status
TBStatus <- TBStatus_temp <- as.character(col_data$TBStatus)
unique(TBStatus_temp)
TBStatus <- ifelse(TBStatus_temp=="LATENT TB","Latent",TBStatus_temp)
col_data$TBStatus <- TBStatus
```

## Create Summazried Experiment Obeject
```{r Summarized experiment object illumina3}
# Create Metadata
GSE19442_experimentData <- new('MIAME',
                      name="Damien Chaussabel",
                      lab="Baylor Institute for Immunology Research",
                      contact="DChaussabel@benaroyaresearch.org",
                      title="An interferon-inducible neutrophil-driven blood transcriptional signature in human tuberculosis",
                      abstract="Whole blood collected in tempus tubes from patients with different spectra of TB disease. All patients were sampled prior to the initiation of any antimycobacterial therapy.",
                      url="10.1038/nature09247",
                      pubMedIds = '20725040',
                      other=list(Platform = 'Illumina HumanHT-12 V3.0 expression beadchip (GPL6947)'))

GSE19442_sobject <- SummarizedExperiment(assays = list(GSE19442_Non_normalized_counts= as.matrix(dat_non_normalized)), colData = col_data, rowData = row_data, metadata = list(GSE19442_experimentData)); GSE19442_sobject

```


# Create Summarized Experiment Object from Affymetrix
## Download in Non-normalized data and confuct RMA
```{r obtain data information affy2.0, cache = TRUE, results="hide"}
geo_access <- "GSE36238"

# clear temporary directory, produce error in Biocstyle when knit
#unlink(paste0(normalizePath(tempdir()), "/", dir(tempdir())), recursive = TRUE) 
GSE36238_normalized_rma <- download_data_Affy(geo_access)

# Rename colnames: Convert GSM1327554.CEL.gz to GSM1327554
colnames(GSE36238_normalized_rma) <- sapply(1:ncol(GSE36238_normalized_rma), 
                                          function(x) gsub('\\..*','',colnames(GSE36238_normalized_rma)[x]))
```

## Create Row Data
```{r create row data affy2.0, results="hide"}
# use function from the package
row_data <- create_RowData(GSE36238_normalized_rma,"GPL570")
dat_download <- GSE36238_normalized_rma
plat_access <- "GPL570"
create_RowData <- function(dat_download, plat_access){

  dat_download <- data.frame(dat_download)
  # Obatian sample names
  samplename <- colnames(dat_download)

  # Read in data from vendor's information
  HumanHT <- GEOquery::getGEO(plat_access, GSEMatrix = F)
  HumanHT_dat <- HumanHT@dataTable@table

  # Annotation
  PROBES <- row.names(dat_download)
  library(illuminaHumanv3.db)
  library(illuminaHumanv4.db)
  library(hgu133plus2.db)
  if (plat_access == "GPL6947"){
    OUT <- AnnotationDbi::select(illuminaHumanv3.db, PROBES, "SYMBOL")
    OUT[is.na(OUT)] <- NA
  }
  if (plat_access== "GPL10558"){
    OUT <- AnnotationDbi::select(illuminaHumanv4.db, PROBES, "SYMBOL")
    OUT[is.na(OUT)] <- NA
  }
  # Affymetrix 2.0
  if (plat_access== "GPL570"){
    OUT <- AnnotationDbi::select(hgu133plus2.db, PROBES, "SYMBOL")
    OUT[is.na(OUT)] <- NA
      # Map ProbeID to Gene Symbol
  OUT_collapse <- OUT %>% dplyr::group_by(PROBEID) %>%
    dplyr::summarise(SYMBOL = paste(SYMBOL, collapse="///"),
                     times = length(unlist(strsplit(SYMBOL, "///"))))
  dat_download$ID_REF <- row.names(dat_download)
  dat_download_final <- dat_download %>% dplyr::left_join(OUT_collapse, by=c('ID_REF' = "PROBEID")) %>% dplyr::left_join(HumanHT_dat, by = c('ID_REF'='ID'))

  # Create Row data annotation
  row_data <- dat_download_final %>% dplyr::select(-samplename)
  Symbol_R <- row_data$SYMBOL
  Symbol_plat <- row_data$`Gene Symbol`

  # Use platform annotation as reference
  Symbol_plat_new <- Symbol_plat
  for (i in 1:length(Symbol_R)){
    if (Symbol_plat_new[i]==""){
      Symbol_plat_new[i] = Symbol_R[i]
    }
  }
  # Create new variable called SYMBOL_NEW, used later in creating multi-assay pbject
  row_data$SYMBOL_NEW <- Symbol_plat_new

  return(DataFrame(row_data))

  }

  # Map ProbeID to Gene Symbol
  OUT_collapse <- OUT %>% dplyr::group_by(PROBEID) %>%
    dplyr::summarise(SYMBOL = paste(SYMBOL, collapse="///"),
                     times = length(unlist(strsplit(SYMBOL, "///"))))
  dat_download$ID_REF <- row.names(dat_download)
  dat_download_final <- dat_download %>% dplyr::left_join(OUT_collapse, by=c('ID_REF' = "PROBEID")) %>% dplyr::left_join(HumanHT_dat, by = c('ID_REF'='ID'))

  # Create Row data annotation
  row_data <- dat_download_final %>% dplyr::select(-samplename)
  Symbol_R <- row_data$SYMBOL
  Symbol_plat <- row_data$Symbol

  # Use platform annotation as reference
  Symbol_plat_new <- Symbol_plat
  for (i in 1:length(Symbol_R)){
    if (Symbol_plat_new[i]==""){
      Symbol_plat_new[i] = Symbol_R[i]
    }
  }
  # Create new variable called SYMBOL_NEW, used later in creating multi-assay pbject
  row_data$SYMBOL_NEW <- Symbol_plat_new

  return(DataFrame(row_data))

}

head(row_data)[1:5,1:5]
```

## Create Column Data
```{r download sample information affy2.0,message = FALSE}
gse <- getGEO(geo_access, GSEMatrix = FALSE)
```

```{r create column information affy2.0}
col_data <- create_ColData(geo_access,gse)
colnames(col_data) <- c('TimePoint','Tissue','Pateient_id','SampleTime')

indx <- sapply(1:length(names(GSMList(gse))), function(i) 
  which(names(GSMList(gse)) %in% colnames(GSE36238_normalized_rma)[i]));indx
col_data <- col_data[indx,]

row.names(col_data) <- colnames(GSE36238_normalized_rma)

col_data_new <- create_new_ColData(col_data)
unique(col_data_new$TBStatus)
col_data_new$TBStatus <- "PTB"
```

## Create Summazried Experiment Obeject
```{r Summarized experiment object affy2.0}
# Create Metadata
# Create Metadata
GSE36238_experimentData <- new('MIAME',
                      name="Jackie Cliff",
                      lab="London School of Hygiene & Tropical Medicine",
                      contact="jackie.cliff@lshtm.ac.uk",
                      title="Distinct phases of blood gene expression pattern through tuberculosis treatment reflect modulation of the humoral immune response.",
                      abstract="An example SummarizedExperimentObject",
                      url="10.1093/infdis/jis499",
                      pubMedIds = '22872737',
                      other=list(Platform = 'Affymetrix Human Genome U133 Plus 2.0 Array: GPL570'))

GSE36238_summarizedExperiemnt <- SummarizedExperiment(assays = list(GSE36238_normalized_counts= as.matrix(GSE36238_normalized_rma)), colData = col_data_new, rowData = row_data, metadata = list(GSE36238_experimentData)); GSE36238_summarizedExperiemnt

```



