---
title: "curatedTBData"
author: "Xutao Wang"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{curatedTBData}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache.lazy = FALSE)
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
if (! "SummarizedExperiment" %in% installed.packages()) BiocManager::install("SummarizedExperiment")
if (! "MultiAssayExperiment" %in% installed.packages()) BiocManager::install("MultiAssayExperiment")
if (! "limma" %in% installed.packages()) install.packages("plyr")
if (! "GEOquery" %in% installed.packages()) BiocManager::install("GEOquery")
if (! "dplyr" %in% installed.packages()) install.packages("dplyr")
if (! "plyr" %in% installed.packages()) install.packages("plyr")
if (! "TBSignatureProfiler" %in% installed.packages()) devtools::install_github("aubreyodom/TBSignatureProfiler")
```

```{r libraries, message = FALSE}
library(GEOquery)
library(SummarizedExperiment)
library(MultiAssayExperiment)
library(limma)
library(edgeR)
library(dplyr)
library(plyr)

library(TBSignatureProfiler)
library(ggplot2)
library(gridExtra)
library(ggridges)

# library(magick)
# Load probe db
library(illuminaHumanv4.db) # Illumina HumanHT-12 V4.0 expression beadchip

devtools::load_all(".")
library(curatedTBData)
```
# `curatedTBData`: Clinically Annotated Data for the Tuberculosis (TB) Cancer Transcriptome
Insert introduction/background infrmation about this data package
# Using SummarizedExperiment/MultiAssayExperiment Object
Insert how you get these objects from the beginnig 
```{r}

```

# Load list of existing SummarizedExperiment/MultiAssayExperiment Objects
We use `SummarizedExperiment` object to store information for microarray datasets, and use `MultiAssayExperiment` object to store information for RNA sequencing datasets
```{r load list of sobject, message=FALSE}
# Load the pacakge
list_files <- data(package="curatedTBData")[["results"]][,"Item"]

# Get a list of all data
data(list=list_files) # Takes some time...
strobject <- ls(pattern="*object$")

# Create a list for loaded data
objects_list <- lapply(strobject, function(x) get(x))
names(objects_list) <- gsub("_.*","",strobject)

rna_seq_name <- which(sapply(objects_list, function(x) class(x)) == "MultiAssayExperiment") %>% names()
microarray_name <- which(sapply(objects_list, function(x) class(x)) != "MultiAssayExperiment") %>% names()
data_type <- data.frame(GSE = c(rna_seq_name,microarray_name), Type = c(rep("RNA-seq",length(rna_seq_name)),rep("Microarray",length(microarray_name))))

```
# Subsetting data with specfic instruction

# Normalization (Before match probeID to Gene symbol)
For imported expression datasets other than affymetrix microarray and GSEXXXXX, they are all non-normalized raw data. Next, we normalize microarray data using `normalizeBetweenArrays` from `limma` package , and normalize RNA-seq data using `TMM` from `edgeR` package.

## Between-sample normalization for Illumina microarray
should create function for this??
```{r}
# add the normalized data as a new experiment in the SummarizedExperiment 
# set all counts less than 0 to be 1
assays(GSE39939_sobject)[["GSE39939_Normalized_counts"]] <- limma::normalizeBetweenArrays (assays(GSE39939_sobject)[[1]],method="quantile")
```

## Betwwen-sample normalization for RNA-seq
Should create a function for this ??
```{r}
counts <- experiments(GSE107991_mobject)[["assay_raw"]] %>% assay() 
TMM <- calcNormFactors(counts, method="RLE")
tmmScaleFactors <- colSums(counts) * TMM
tmmExp <- round(t(t(counts)/tmmScaleFactors) * mean(tmmScaleFactors)) 
# add the normalized data as a new experiment in the MultiAssay 
k=paste0("a","aaaas")

GSE107991_mobject_norm <- c(GSE107991_mobject,k = tmmExp) 

```


# Match ProbeID to Gene symbol
For objects that contain probe information, we match them to gene symbol. We collapse probe information by taking their mean values if they mapped to the same gene, and expand probe information for non-uniqueness mapping. In such cases, we use MultiAssayExperiment Objects to store information for each data.
```{r}
# Create MultiAssayExperiment Object from SummarizedExperiment Object
GSE39939_mobject <- Create_MultiAssay_object(GSE39939_sobject)

# Create MultiAssayExperiment Object from MultiAssayExperiment Object
GSE107991_mobject_new <- Create_MultiAssay_object(GSE107991_mobject)
```

```{r}
# Match ProbeID to Gene symbol for all existing datasets
multi_data_all <- lapply(objects_list, function(x) Create_MultiAssay_object(x))
```

# TBSignature Analysis Using `TBSignatureProfiler`{.tabset}
```{r}

```

## Positive TB (PTB) vs. Latent TB (Latent){.tabset}
```{r calculate ssgsea for PTB vs Latent, message = FALSE, results = "hide"}
multi_set_PTB_Latent <- lapply(multi_data_all, function(x) get_sobject_TBSig(x,"PTB","Latent")) %>% plyr::compact()

# 31 datasets contain information for PTB vs. Latent
ssgsea_PTB_Latent <- lapply(multi_set_PTB_Latent,
                    function(x) TBSignatureProfiler::runTBsigProfiler(input = x,
                                  useAssay = "counts",
                                  signatures = TBsignatures,
                                  algorithm = "ssGSEA",
                                  combineSigAndAlgorithm = TRUE,
                                  parallel.sz = 12))

# No Sloot_HIV_2, Suliman_RISK_4 in some datasets
TBsig_PTB_Latent <- names(TBsignatures)[-grep("Sloot_HIV_2|Suliman_RISK_4",names(TBsignatures))]

index_Sloot <- which(!sapply(ssgsea_PTB_Latent, function(x) "Sloot_HIV_2" %in% names(colData(x))))
index_Suliman <- which(!sapply(ssgsea_PTB_Latent, function(x) "Suliman_RISK_4" %in% names(colData(x))))
# Define new function for box plot display
get_boxplot <- function(result_list,sig_name){
  
  p_boxplot <- lapply(1:length(result_list), function(x)
    TBSignatureProfiler::signatureBoxplot(inputData = result_list[[x]],
                                      name = names(result_list)[x],
                                      signatureColNames = sig_name,
                                      annotationColName = "Disease", rotateLabels = FALSE,fill_colors = c("#4E84C4", "#FC4E07")))

    library(gridExtra)
    library(ggplot2)
    p_combine <- do.call("grid.arrange", c(p_boxplot, ncol=floor(sqrt(length(p_boxplot)))))
    return(p_combine)

}
```

### Boxplot of all scores signatures{.tabset}

```{r Boxplot for each signature PTB vs. Latent,results = "asis", message = FALSE, fig.height = 20, fig.width=28}
for (i in TBsig_PTB_Latent){

  cat("####", i, "\n")
  
  get_boxplot(ssgsea_PTB_Latent, sig_name = i)
  #print(signatureBoxplot(ssgsea_result, name = i, signatureColNames = i,
  #               annotationColName = c("Disease")))
  
  cat("\n\n")
}
```

#### Sloot_HIV_2

```{r boxplot for Sloot_HIV_2 PTB vs. Latent, results="hide",fig.height = 16, fig.width=24}
get_boxplot(ssgsea_PTB_Latent[-index_Sloot], sig_name = "Sloot_HIV_2")

```

#### Suliman_RISK_4

```{r boxplot for Suliman_RISK_4 PTB vs. Latent, results="hide",fig.height = 16, fig.width=24}
get_boxplot(ssgsea_PTB_Latent[-index_Suliman], sig_name = "Suliman_RISK_4")

```

# Train predictors using ensemble 
## Batch Correction

