---
title: "curatedTBData"
author: "Xutao Wang"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{curatedTBData}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache.lazy = FALSE)
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
if (! "BiocParallel" %in% installed.packages()) BiocManager::install("BiocParallel")
if (! "SummarizedExperiment" %in% installed.packages()) BiocManager::install("SummarizedExperiment")
if (! "MultiAssayExperiment" %in% installed.packages()) BiocManager::install("MultiAssayExperiment")
if (! "limma" %in% installed.packages()) install.packages("plyr")
if (! "GEOquery" %in% installed.packages()) BiocManager::install("GEOquery")
if (! "dplyr" %in% installed.packages()) install.packages("dplyr")
if (! "plyr" %in% installed.packages()) install.packages("plyr")
if (! "TBSignatureProfiler" %in% installed.packages()) devtools::install_github("aubreyodom/TBSignatureProfiler")
```

```{r libraries, message=FALSE}
# Alphabetic order
library(BiocParallel)
library(dplyr)
library(edgeR)
library(GEOquery)
library(ggplot2)
library(ggridges)
library(gridExtra)
library(illuminaHumanv4.db) # Illumina HumanHT-12 V4.0 expression beadchip
library(limma)
library(MultiAssayExperiment)
library(plyr)
library(SummarizedExperiment)
library(sva)
library(TBSignatureProfiler)


# library(magick)
# Load probe db

devtools::load_all(".")
library(curatedTBData)
```

# `curatedTBData`: Clinically Annotated Data for the Tuberculosis (TB) Transcriptome

Insert introduction/background infrmation about this data package

# Using SummarizedExperiment/MultiAssayExperiment Object

## Create SummarizedExperiment Object from Affymetrix 2.0 Microarray

```{r create Affy2.0, message=FALSE, warning=FALSE}
# GSE36238
GSE36238_raw <- curatedTBData(geo_access = "GSE36238", plat_access = "GPL570")

# Enter clinical information...
colnames(GSE36238_raw$col_data) <- c('TimePoint','Tissue','Patient_id','SampleTime')
GSE36238_raw$col_data$TBStatus <- "PTB"

# Convert to standard column feature
GSE36238_raw$col_data_new <- create_standard_ColData(GSE36238_raw$col_data)

# Obtain Metadata
GSE36238_experimentData <- new('MIAME',
                      name="Jackie Cliff",
                      lab="London School of Hygiene & Tropical Medicine",
                      contact="jackie.cliff@lshtm.ac.uk",
                      title="Distinct phases of blood gene expression pattern through tuberculosis treatment reflect modulation of the humoral immune response.",
                      abstract="An example SummarizedExperimentObject",
                      url="10.1093/infdis/jis499",
                      pubMedIds = '22872737',
                      other=list(Platform = 'Affymetrix Human Genome U133 Plus 2.0 Array: GPL570'))

GSE36238 <- SummarizedExperiment(assays = list(GSE36238_normalized_counts= as.matrix(GSE36238_raw$counts)), colData = GSE36238_raw$col_data_new, rowData = GSE36238_raw$row_data, metadata = list(GSE36238_experimentData)); GSE36238
```

## Create SummarizedExperiment Object from Illumina 3.0 Micoraray

```{r create Illumina 3.0, message=FALSE, warning=FALSE}
# GSE19442
GSE19442_raw <- curatedTBData("GSE19442","GPL6947")

# Enter clinical information...
colnames(GSE19442_raw$col_data)  <- c('Age','Gender','Ethnicity','TBStatus','GeographicalRegion','BcgVaccinated','BirthRegion','TST', 'SputumSmearStatus','sputum_culture')

# Rename TB Status
TBStatus <- TBStatus_temp <- as.character(GSE19442_raw$col_data$TBStatus)
unique(TBStatus_temp)
TBStatus <- ifelse(TBStatus_temp=="LATENT TB","Latent",TBStatus_temp)
GSE19442_raw$col_data$TBStatus <- TBStatus

# Convert to standard column feature
GSE19442_raw$col_data_new <- create_standard_ColData(GSE19442_raw$col_data)

# Obtain Metadata
GSE19442_experimentData <- new('MIAME',
                      name="Damien Chaussabel",
                      lab="Baylor Institute for Immunology Research",
                      contact="DChaussabel@benaroyaresearch.org",
                      title="An interferon-inducible neutrophil-driven blood transcriptional signature in human tuberculosis",
                      abstract="Whole blood collected in tempus tubes from patients with different spectra of TB disease. All patients were sampled prior to the initiation of any antimycobacterial therapy.",
                      url="10.1038/nature09247",
                      pubMedIds = '20725040',
                      other=list(Platform = 'Illumina HumanHT-12 V3.0 expression beadchip (GPL6947)'))

# Create SummarizedExperiment Object for GSE19442
GSE19442 <- SummarizedExperiment(assays = list(GSE19442_Non_normalized_counts= as.matrix(GSE19442_raw$counts)), colData = GSE19442_raw$col_data_new, rowData = GSE19442_raw$row_data, metadata = list(GSE19442_experimentData)); GSE19442
```

## Create SummarizedExperiment Object from Illumina 4.0 Micoraray

```{r}
# GSE39939 to be cotinued
```

# Load available SummarizedExperiment/MultiAssayExperiment Objects

We use `SummarizedExperiment` object to store information for microarray datasets, and use `MultiAssayExperiment` object to store information for RNA sequencing datasets

```{r load list of sobject, message=FALSE}
# Load the pacakge
list_files <- data(package="curatedTBData")[["results"]][,"Item"]

# Get a list of all data
data(list=list_files) # Takes some time...
strobject <- ls(pattern="*object$")

# Create a list for loaded data for metadata analysis
objects_list <- lapply(strobject, function(x) get(x))
names(objects_list) <- gsub("_.*","",strobject)

#rna_seq_name <- which(sapply(objects_list, function(x) class(x)) == "MultiAssayExperiment") %>% names()
#microarray_name <- which(sapply(objects_list, function(x) class(x)) != "MultiAssayExperiment") %>% names()
#data_type <- data.frame(GSE = c(rna_seq_name,microarray_name), Type = c(rep("RNA-seq",length(rna_seq_name)),rep("Microarray",length(microarray_name))))

```

# Subsetting Objects

The major adavantage of using `SummarizedExperiment` and `MultiAssayExperiment` is the coordination of the meta-data and assays when subsetting.

## Subsetting for single objects

Both objects have similar method used for subsetting column information. 
```{r}
# subsetting samples from objects with TBStatus: PTB
GSE39939_sobject[, GSE39939_sobject$TBStatus == "PTB"]
```

```{r}
# subsetting samples from objects with TBStatus: PTB
GSE107991_mobject[, GSE107991_mobject$TBStatus == "PTB"]
```

## Subsetting for multiple objects based on multiple conditions

Since not all datasets contain information for each TB status. Next we want to select out objects that contain samples with only positive TB (PTB) and latent TB (Latent).
```{r, warning=FALSE}
objects_list1 <-  lapply(objects_list, function(x,disease1,disease2) {
  x_filter <- x[,x$TBStatus %in% c(disease1,disease2)]
  TB_status <- SummarizedExperiment::colData(x_filter)["TBStatus"][,1]
  if(length(unique(TB_status)) == 2){
    return(x_filter)
  }
},"PTB","Latent") %>% plyr::compact()

# 21/42 objects contain samples from PTB and Latent
length(objects_list1) 

```


# Normalization

For imported expression datasets other than affymetrix microarray and GSEXXXXX, they are all non-normalized raw data. Next, we normalize microarray data using `normalizeBetweenArrays` from `limma` package , and normalize RNA-seq data using `TMM` from `edgeR` package before matching probesets to gen symbol. 
Steps take: 
1. remove outliers using `arrayQualityMetrics`. 
2. Intensity values < 10 set to 10
3. log2 transformed the data.
4. Normalize data.

## Between-sample normalization for Illumina microarray

```{r}
# add the normalized data as a new assay in the SummarizedExperiment 
# set all counts less than 0 to be 1
GSE39939_sobject_norm <- NormalizeReads(GSE39939_sobject, method = "quantile")
```
Boxplot for non-normalized microarray data
```{r, fig.height = 8, fig.width = 8}
boxplot(log(assays(GSE39939_sobject_norm)[["GSE39939_Non_normalized_counts"]][,1:20], base=2))
# display the boxplot in log2 for selected samples
```
Boxplot for microarray data after quantile normalization
```{r, fig.height = 8, fig.width = 8}
boxplot(assays(GSE39939_sobject_norm)[["NormalizedData"]][,1:20])
# display the boxplot in log2 for selected samples
```

## Betwwen-sample normalization for RNA-seq

```{r}
# Add the normalized data as a new assay in the MultiAssayExperiment. 
# Use experiment_type = "assay_reprocess" for reprocess RNA-seq
GSE107991_mobject_norm <- NormalizeReads(GSE107991_mobject, experiment_type = "assay_raw", method = "TMM")
```

Boxplot for non-normalized RNA-seq data
```{r, fig.height = 8, fig.width = 8}
# display the boxplot in log2
boxplot(log(assay(experiments(GSE107991_mobject_norm)[["assay_raw"]])[,1:20]+0.0001, base = 2))
```

Boxplot for normalized RNA-seq data
```{r, fig.height = 8, fig.width = 8}
# display the boxplot in log2
# TMM reduces maximum value
boxplot(assays(experiments(GSE107991_mobject_norm)[["assay_raw"]])[["NormalizedData"]][,1:20])
```

# Match ProbeID to Gene symbol

For objects that contain probe information, we match them to gene symbol. We collapse probe information by taking their mean values if they mapped to the same gene, and expand probe information for non-uniqueness mapping. In such cases, we use MultiAssayExperiment Objects to store information for each data.

```{r}
# Match Probe ID using raw data
# Specificy experiment_type = "assay_raw_norm" fo matching probe ID using normalized data (if applicable)
MatchProbe(GSE39939_sobject, experiment_type = "assay_raw")
MatchProbe(GSE107991_mobject, experiment_type = "assay_raw")
```

# TBSignature Analysis 
First of all, we create reate a list of object with normalized reads and probesets mapped to symbols. Match probe ID to symbol using either raw data or normalized data (if applicable) .
Matching probes to symbol using raw data is straight-forward, we use the normalized data in the following.

```{r, message=FALSE}
# Normalization. For RNA-seq data, we use raw data from online for normalization (Not the self-processed data)
# Use online raw data by specifying the experiment_type=assay_raw
object_norm <- lapply(objects_list, function(x) NormalizeReads(x, experiment_type = "assay_raw")) 

# Match Probe to gene symbols
# Use the normalized data from assay_raw by specifying the experiment_type=assay_raw_norm
object_match <- lapply(object_norm, function(x) MatchProbe(x, experiment_type = "assay_raw_norm"))
```

## Active TB (PTB) vs. Latent TB (Latent) using `TBSignatureprofiler`{.tabset}
`TBSignatureprofiler` is a package to evaluate the the performance of signatures across curated TB data. We subset samples with PTB and Latent TB status for binary classification using function `SubsetDisease`, and implement Single-sample Gene Set Enrichment Analysis (ssGSEA) scoring algorithm to compute gene scores for each signature.
Workflow of ssGSEA:
1. Input data, take sample-specific absolute rank.
2. After getting the ranked data, calculate random walk statistics for each sample. p: Number of genes, j: sample, k:gene set
$$v_{jk}(\ell)=\frac{\sum_{i=1}^\ell r_{ij}^\tau I(g(i)\in\gamma_k)}{\sum_{i=1}^pr_{ij}^\tau I(g(i)\in\gamma_k)}-\frac{\sum_{i=1}^\ell I(g(i)\notin \gamma_k)}{p-|\gamma_k|}, \quad\ell=1,\dots,p \\
ES_{jk}=\sum_{\ell=1}^p v_{jk}(\ell)$$
3. Re-scale the enrichment scores (ES) for each sample
$$\frac{ES_{jk}}{\max(ES_{jk})-\min(ES_{jk})}, j=1,\dots,n$$
```{r calculate ssgsea for PTB vs Latent, message = FALSE, warning = FALSE, results = "hide"}
# Select sample contain information about PTB and Latent
# Modify the list to make it compatible with TBSignatureProfiler

multi_set_PTB_Latent <- lapply(object_match, function(x)
  SubsetDisease(x,c("Latent","PTB"), experiment_type="assay_reduce")) %>% plyr::compact() 
# Remove empty objects use plyr::compact()

ssgsea_PTB_Latent <- lapply(multi_set_PTB_Latent,
                    function(x) TBSignatureProfiler::runTBsigProfiler(input = x,
                                  useAssay = assayNames(x),
                                  signatures = TBsignatures, # list of Signatures from TBSignatureProfiler
                                  algorithm = "ssGSEA",
                                  combineSigAndAlgorithm = TRUE,
                                  parallel.sz = 12))

# Get signatures with only scores and disease status
# TBSignature, a list of desired signarure signature 
ssgsea_PTB_Latent1 <- SignatureFilter(ssgsea_PTB_Latent, TBsignatures)

# No Sloot_HIV_2, Suliman_RISK_4 in some datasets
TBsig_PTB_Latent <- names(TBsignatures)[-grep("Sloot_HIV_2|Suliman_RISK_4",names(TBsignatures))]

index_Sloot <- which(!sapply(ssgsea_PTB_Latent, function(x) "Sloot_HIV_2" %in% names(colData(x))))
index_Suliman <- which(!sapply(ssgsea_PTB_Latent, function(x) "Suliman_RISK_4" %in% names(colData(x))))
```

### Boxplot of all scores signatures{.tabset}

```{r Boxplot for each signature PTB vs. Latent,results = "asis", message = FALSE, fig.height = 25, fig.width=28}
for (i in names(TBsignatures)){

  cat("####", i, "\n")
  
  Boxplot_TBSig(ssgsea_PTB_Latent1, sig_name = i)
  #print(signatureBoxplot(ssgsea_result, name = i, signatureColNames = i,
  #               annotationColName = c("Disease")))
  
  cat("\n\n")
}
```

### Comparing Signature Scores via AUCs

```{r AUC ridge plot for PTB vs Latent,fig.height = 15, fig.width=8, message=FALSE}
# Combine results from ssgsea_PTB_Latent
ssgsea_PTB_Latent_combine <- combine_auc(ssgsea_PTB_Latent, TBsignatures)

# Ridge plot for AUCs distribution across datasets
# PTB vs. Latent
get_auc_distribution(ssgsea_PTB_Latent_combine) + ggtitle("Ridge plot of AUC for PTB vs. Latent")
```

## Molecular Degree of Perturbation Analysis
Molecular Degree of Perturbation (MDP) is a method to standardized gene expression data to the refernce (control). First of all, we subset the lists that contain samples with TBStatus: Control.
```{r MDP, message=FALSE}
MDP_control_name <- lapply(object_match, function(x) 
  x[,x$TBStatus == "Control"]) %>% remove_empty_object() %>% names()
MDP_control<-  object_match[MDP_control_name] # study includes control samples

# -c(9,19,21)

# Temporary result
MDP_result <- lapply(1:8, function(x) MDP(MDP_control[[x]], gset = TBsignatures, experiment_type = "assay_reduce", assay_name = 1, GSE = names(MDP_control)[x]))
```

### Boxplot for MDP{.tabset}

```{r, results = "asis", message = FALSE, fig.height = 20, fig.width=28}
for (i in names(TBsignatures)){

  cat("####", i, "\n")
  
  Boxplot_TBSig(MDP_result, sig_name = i)
  #print(signatureBoxplot(ssgsea_result, name = i, signatureColNames = i,
  #               annotationColName = c("Disease")))
  
  cat("\n\n")
}
```


# Integration of RNA-Seq data with microarray data for TB profiling

## Batch Correction with ComBat

If datasets are merged, it is typically recommended to remove a very likely batch effect. We will use the ComBat method, implemented for example in the SVA Bioconductor package. Start using `CombineObjects` to intersect objects with common gene symbol. This function produces a SummarizedExperiment object.
```{r, message=FALSE}
# Select the objects you want to combine
# Randomly pick 2 from RNA-seq,2 from microaray
result <- CombineObjects(object_match, c("GSE101705","GSE107104", "GSE54992", "GSE19444")) 
# output is a SummarizedExperiment Object

# Batch Correction
mod1 <- model.matrix(~as.factor(TBStatus), data=colData(result))
batch1 <- colData(result)$GSE
combat_edata1 <- sva::ComBat(dat=assays(result)[[1]], batch=batch1, mod=mod1)

# Add batch corrected data as a new assay to the orginal SummarizedExperiment
assays(result)[["BacthCorrection"]] <- combat_edata1
```
Boxplot before Batch correction
```{r, fig.width=8, fig.height=6, warning=FALSE}
boxplot(assays(result)[[1]])
```

Boxplot after Batch correction
```{r, fig.width=8, fig.height=6, warning=FALSE}
# log 2
boxplot(assays(result)[[2]])
```

## Molecular Degree of Perturbation (MDP)
