---
title: "MDP_integration"
author: "Xutao Wang"
date: "5/9/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, cache.lazy = FALSE, message=FALSE, warning=FALSE)
```

```{r install curatedTBData, eval = FALSE}
devtools::install_github("xutao-wang/curatedTBData")
```

```{r libraries, message=FALSE}
library(dplyr)
library(edgeR)
library(GEOquery)
library(ggplot2)
library(ggridges)
library(gridExtra)
library(limma)
library(MultiAssayExperiment)
library(plyr)
library(SummarizedExperiment)
library(sva)
library(TBSignatureProfiler)

devtools::load_all()
library(curatedTBData)
```


# curatedTBData
## Import Data
```{r load a list of datasets}
# To get a signle dataset, run the following:
# data(package="curatedTBData")
# data(GSE39939_sobject)

list_files <- data(package="curatedTBData")[["results"]][,"Item"]

# Get a list of all data
data(list=list_files) # Takes some time...
strobject <- ls(pattern="*object$")

# Create a list for loaded data for metadata analysis
objects_list <- lapply(strobject, function(x) get(x)) # Takes some time...
objects_name <- gsub("_.*","",strobject)
names(objects_list) <- objects_name
```

## Normalization/Matching
```{r normalized/matched sample, cache=TRUE}
# Normalization. For RNA-seq data, we use raw data from online for normalization (Not the self-processed data)
# Use online raw data by specifying the experiment_type=assay_raw
object_norm <- lapply(objects_list, function(x) Normalization(x, experiment_type = "assay_raw")) 

# Match Probe to gene symbols
# Use the normalized data from assay_raw by specifying the experiment_type=assay_raw_norm
object_match1 <- lapply(object_norm, function(x) MatchProbe(x, experiment_type = "assay_raw_norm"))

# Remove samples with TBStatus == NA
object_match <- lapply(object_match1, function(x) x[,colData(x)[,"TBStatus"]!= "NA"])
```

## Molecular Degree of Perturbation Analysis{.tabset}

### Calculate MDP scores

Molecular Degree of Perturbation (MDP) is a method to standardized gene expression data to the refernce (control). First of all, we subset the lists that contain samples with TBStatus: Control.
```{r MDP, cache=TRUE}
MDP_control_name <- lapply(object_match, function(x) 
  x[,x$TBStatus == "Control"]) %>% remove_empty_object() %>% names()
MDP_control <- object_match[MDP_control_name] # study includes control samples

MDP_result <- lapply(1:length(MDP_control), function(x) MDP(MDP_control[[x]], gset = TBsignatures, experiment_type = "assay_reduce", assay_name = 1, GSE = names(MDP_control)[x]))

# Use the original methods by selecting top 25% genes
MDP_result_NULL <- lapply(1:length(MDP_control), function(x) MDP(MDP_control[[x]], gset = NULL, experiment_type = "assay_reduce", assay_name = 1, GSE = names(MDP_control)[x]))
```

### Boxplot for MDP{.tabset}

```{r, results = "asis", message = FALSE, fig.height = 30, fig.width=28, fig.wide=TRUE}
for (i in names(TBsignatures)){

  cat("####", i, "\n")
  
  BoxplotTBSig(MDP_result, gset = i, annotationColName = "TBStatus")
  
  cat("\n\n")
}
```

#### Default 

```{r, results = "asis", message = FALSE, warning=FALSE, fig.height = 30, fig.width=28, fig.wide=TRUE}
BoxplotTBSig(MDP_result_NULL, gset = "sMDP", annotationColName = "TBStatus")
# sMDP is the annotation column name from MDP_result_NULL
```

# Integration of RNA-Seq data with microarray data for TB profiling

## Batch Correction with ComBat

If datasets are merged, it is typically recommended to remove a very likely batch effect. We will use the ComBat method, implemented for example in the SVA Bioconductor package. Start using `CombineObjects` to intersect objects with common gene symbol. This function produces a SummarizedExperiment object.
```{r, message=FALSE}
# Select the objects you want to combine
# Randomly pick 2 from RNA-seq,2 from microarray
result <- CombineObjects(object_match, gse_name=c("GSE101705","GSE107104", "GSE54992", "GSE19444"), experiment_type="assay_reduce") 
# output is a SummarizedExperiment Object

# Batch Correction
mod1 <- model.matrix(~as.factor(TBStatus), data=colData(result))
batch1 <- colData(result)$GSE
combat_edata1 <- sva::ComBat(dat=assays(result)[[1]], batch=batch1, mod=mod1)

# Add batch corrected data as a new assay to the orginal SummarizedExperiment
assays(result)[["BacthCorrect"]] <- combat_edata1
```

Boxplot before Batch correction

```{r, fig.width=8, fig.height=6, warning=FALSE}
boxplot(assays(result)[[1]])
```

Boxplot after Batch correction
```{r, fig.width=8, fig.height=6, warning=FALSE}
# log 2
boxplot(assays(result)[[2]])
```

## Molecular Degree of Perturbation (MDP) for Batch Corrected Data
Recall the list of studies that have control samples. We could either pick up some data from the data for batch correction or use all of them. Next we will use all the available 21 studies. Function `CombineObjects` enables us to combine studies from different sources/platforms by intersecting commone genes from each dataset. The output data can be used for batch correction.
```{r Batch corrected of MDP/ssGSEA, cache=TRUE, message = FALSE, warning = FALSE, results = "hide"}
# Batch correction
# remove GSE74092 becuase it only has 379 rows (Some are NA's)
# because GSE74092 used quantitative PCR so they did not have enough coverage to capture all genes
MDP_batch <- CombineObjects(object_match, gse_name = MDP_control_name[-which(MDP_control_name %in% "GSE74092")], experiment_type = "assay_reduce") 

# Batch Correction
mod1 <- model.matrix(~as.factor(TBStatus), data=colData(MDP_batch))
batch1 <- colData(MDP_batch)$GSE
combat_edata1 <- sva::ComBat(dat=assays(MDP_batch)[[1]], batch=batch1, mod=mod1)

# Add batch corrected data as a new assay to the orginal SummarizedExperiment
assays(MDP_batch)[["BacthCorrect"]] <- combat_edata1

# Calculate MDP for batch correct object
MDP_batch_result <- MDP(theObject = MDP_batch, gset = TBsignatures, assay_name = "BacthCorrect")

# Calculate ssGSE for batch correct object
MDP_batch_ssgsea <- TBSignatureProfiler::runTBsigProfiler(input = MDP_batch,
                                  useAssay = "BacthCorrect",
                                  signatures = TBsignatures, 
                                  algorithm = "ssGSEA",
                                  combineSigAndAlgorithm = TRUE,
                                  parallel.sz = 12)
```

### Boxplot for Batch Corrected MDP

```{r, results = "asis", message = FALSE, warning = FALSE, fig.wide=TRUE, fig.height = 32, fig.width=32}
# 5 levels of TBStatus: Control/Laent/OD/PTB
# Create SummarizedExperiment Object to make it compaticble with the function

BoxplotTBSig(MDP_batch_result, gset = names(TBsignatures), annotationColName = "TBStatus")
```

### Boxplot for batch corrected ssGSEA
```{r, results = "asis", message = FALSE, warning = FALSE, fig.wide=TRUE, fig.height = 32, fig.width=32}
BoxplotTBSig(data.frame(colData(MDP_batch_ssgsea)), gset = names(TBsignatures), annotationColName = "TBStatus")
```

```{r session info}
sessionInfo()
```

