---
title: "curatedTBData"
author: "Xutao Wang"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{curatedTBData}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, cache.lazy = FALSE, message=FALSE, warning=FALSE)
```

```{r install curatedTBData, eval = FALSE}
devtools::install_github("xutao-wang/curatedTBData")
```

```{r libraries, message=FALSE}
library(dplyr)
library(edgeR)
library(GEOquery)
library(ggplot2)
library(ggridges)
library(gridExtra)
library(limma)
library(MultiAssayExperiment)
library(plyr)
library(SummarizedExperiment)
library(sva)
library(TBSignatureProfiler)

devtools::load_all()
library(curatedTBData)
```

This package requires large memory to process data. Please increase R script memory before running this vignette. 
[Increase memory for Macs](https://stackoverflow.com/questions/51295402/r-on-macos-error-vector-memory-exhausted-limit-reached)
Step 1: Open terminal
Step 2:
```{r, engine='bash', eval=FALSE}
cd ~
touch .Renviron
open .Renviron
```
Step 3: Save the following as the first line of .Renviron:
```{r, engine='bash', eval=FALSE}
R_MAX_VSIZE=50Gb 
```

# `curatedTBData`: Clinically Annotated Data for the Tuberculosis (TB) Transcriptome

Insert introduction/background infrmation about this data package

# Create SummarizedExperiment/MultiAssayExperiment Object

# Load available SummarizedExperiment/MultiAssayExperiment Objects

We use `SummarizedExperiment` object to store information for microarray datasets, and use `MultiAssayExperiment` object to store information for RNA sequencing datasets

```{r load a list of datasets}
# To get a signle dataset, run the following:
# data(package="curatedTBData")
# data(GSE39939_sobject)

list_files <- data(package="curatedTBData")[["results"]][,"Item"]

# Get a list of all data
data(list=list_files) # Takes some time...
strobject <- ls(pattern="*object$")

# Create a list for loaded data for metadata analysis
objects_list <- lapply(strobject, function(x) get(x)) # Takes some time...
objects_name <- gsub("_.*","",strobject)
names(objects_list) <- objects_name

#####################################################
## Create summary table for data with their sequencing platform
rna_seq_name <- which(sapply(objects_list, function(x) class(x)) == "MultiAssayExperiment") %>% names()
microarray_name <- which(sapply(objects_list, function(x) class(x)) != "MultiAssayExperiment") %>% names()
data_type <- data.frame(GSE = c(rna_seq_name,microarray_name), Type = c(rep("RNA-seq",length(rna_seq_name)),rep("Microarray",length(microarray_name))), stringsAsFactors = F)
index_norm <- which(data_type$GSE %in% c("GSE54992","GSE36238","GSE31348","GSE73408","GSE41055" ,"GSEXXXXX"))

data_type[index_norm,"Type"] <- "NormalizedMicroarray"

```

# Subsetting Objects

The major adavantage of using `SummarizedExperiment` and `MultiAssayExperiment` is the coordination of the meta-data and assays when subsetting.

## Subsetting for single objects

Both objects have similar method used for subsetting column information. For example, we want to select samples with only active TB
```{r subsetting single object example}
# subsetting samples from objects with TBStatus: PTB
GSE39939_sobject[, GSE39939_sobject$TBStatus == "PTB"]
# subsetting samples from objects with TBStatus: PTB
GSE107991_mobject[, GSE107991_mobject$TBStatus == "PTB"]
```

## Subsetting for multiple objects based on multiple conditions

Since not all datasets contain information for each TB status. Next we want to select out objects that contain samples with only positive TB (PTB) and latent TB (Latent). In this case, we can use `SubsetSample` function from curatedTBData. For either a list of datasets or a single dataset. 
```{r subsetting multiple objects example}
objects_list_test <-  lapply(objects_list, function(x)
  SubsetTBStatus(x,"TBStatus",c("Latent","PTB"), experiment_type = "all")) %>% plyr::compact() 

# 21/42 objects contain samples from PTB and Latent
length(objects_list_test) 
```

# Normalization

For imported expression datasets other than affymetrix microarray and GSEXXXXX, they are all non-normalized raw data. Next, we normalize microarray data using `normalizeBetweenArrays` from `limma` package , and normalize RNA-seq data using `TMM` from `edgeR` package before matching probesets to gen symbol. 
Steps take: 
1. remove outliers using `arrayQualityMetrics`.??? Have not included yet
2. Set intensity values < 10 set to 10
3. log2 transformed the data.
4. Normalize data (RMA for Affymetrix microarray, quantile normalization for other microarray).

## Between-sample normalization for Illumina microarray

```{r}
# add the normalized data as a new assay in the SummarizedExperiment 
# set all counts less than 0 to be 1
GSE39939_sobject_norm <- Normalization(GSE39939_sobject, method = "quantile")
```
Boxplot for non-normalized microarray data
```{r, fig.width = 8, fig.height = 8}
boxplot(log(assays(GSE39939_sobject_norm)[["GSE39939_Non_normalized_counts"]][,1:20], base=2))
# display the boxplot in log2 for selected samples
```
Boxplot for microarray data after quantile normalization
```{r, fig.width = 8, fig.height = 8}
boxplot(assays(GSE39939_sobject_norm)[["NormalizedData"]][,1:20])
# display the boxplot in log2 for selected samples
```

## Betwwen-sample normalization for RNA-seq

```{r}
# Add the normalized data as a new assay in the MultiAssayExperiment. 
# Use experiment_type = "assay_reprocess" for reprocess RNA-seq
GSE107991_mobject_norm <- Normalization(GSE107991_mobject, experiment_type = "assay_raw", method = "TMM")
```

Boxplot for non-normalized RNA-seq data
```{r, fig.width = 8, fig.height = 8}
# display the boxplot in log2
boxplot(log(assay(experiments(GSE107991_mobject_norm)[["assay_raw"]])[,1:20]+0.0001, base = 2))
```

Boxplot for normalized RNA-seq data
```{r, fig.width = 8, fig.height = 8}
# display the boxplot in log2
# TMM reduces maximum value
boxplot(assays(experiments(GSE107991_mobject_norm)[["assay_raw"]])[["NormalizedData"]][,1:20])
```

# Match ProbeID to Gene symbol

For objects that contain probe information, we match them to gene symbol. We collapse probe information by taking their mean values if they mapped to the same gene, and expand probe information for non-uniqueness mapping. In such cases, we use MultiAssayExperiment Objects to store information for each data.

```{r}
# Match Probe ID using raw data
# Specificy experiment_type = "assay_raw_norm" for matching probe ID using normalized data (if applicable)
MatchProbe(GSE39939_sobject, experiment_type = "assay_raw")
MatchProbe(GSE107991_mobject, experiment_type = "assay_raw")
```

# TB Signature Analysis
First of all, we create reate a list of object with normalized reads and probesets mapped to symbols. Match probe ID to symbol using either raw data or normalized data (if applicable) .
Matching probes to symbol using raw data is straight-forward, we use the normalized data in the following.

```{r normalized/matched sample, cache=TRUE}
# Normalization. For RNA-seq data, we use raw data from online for normalization (Not the self-processed data)
# Use online raw data by specifying the experiment_type=assay_raw
object_norm <- lapply(objects_list, function(x) Normalization(x, experiment_type = "assay_raw")) 

# Match Probe to gene symbols
# Use the normalized data from assay_raw by specifying the experiment_type=assay_raw_norm
object_match1 <- lapply(object_norm, function(x) MatchProbe(x, experiment_type = "assay_raw_norm"))

# Remove samples with TBStatus == NA
object_match <- lapply(object_match1, function(x) x[,colData(x)[,"TBStatus"]!= "NA"])

# Add new list of signatures
TBsignatures <- TBSignatureProfiler::TBsignatures
TBsignatures$"Suliman_RISK_4" <- c("GAS6","SEPT4","CD1C","BLK")
TBsignatures$"Suliman_4" <- c("ANKRD22", "C1QC", "OSBPL10", "TRAV27") 
TBsignatures$"Leong_RISK_29"<- c("SRBD1", "ZNF419", "SH2D1B", "CTSA", "GSTA4", "AGAP9", "MOB3C", "WARS", "LUC7L", "ZNRF1", "CIRBP", "PRSS53", "APOL6", "TCN2", "MDN1", "SNRNP70", "SLC3A1", "NAGA", "SPDYE5", "SPSB1", "CCDC14", "IL31RA", "DERA", "FUT4", "NEIL1", "ENO3", "CCDC78", "HM13", "ZNF202")
```

## Active TB (PTB) vs. Latent TB (Latent) using `TBSignatureprofiler`{.tabset}

### Caculate ssGSEA

`TBSignatureprofiler` is a package to evaluate the the performance of signatures across curated TB data. We subset samples with PTB and Latent TB status for binary classification using function `SubsetTBStatus`, and implement single-sample Gene Set Enrichment Analysis (ssGSEA) scoring algorithm to compute gene scores for each signature.
Workflow of ssGSEA:
1. Input data, take sample-specific absolute rank:$|r_{ij}|$.

2. After getting the ranked data, calculate random walk statistics for each sample. p: Number of genes, j: sample, k:gene set
$$v_{jk}(\ell)=\frac{\sum_{i=1}^\ell |r_{ij}|^\tau I(g(i)\in\gamma_k)}{\sum_{i=1}^p|r_{ij}|^\tau I(g(i)\in\gamma_k)}-\frac{\sum_{i=1}^\ell I(g(i)\notin \gamma_k)}{p-|\gamma_k|}, \quad\ell=1,\dots,p \\
ES_{jk}=\sum_{\ell=1}^p v_{jk}(\ell)$$
$p=$ Number of genes.
$|\gamma_k|$= Number of genes within gene set $k$.
$\tau=0.25$.

3. Re-scale the enrichment scores (ES) for each sample
$$ES_{jk}=\frac{ES_{jk}}{\max\{ES_{1k},\dots,ES_{nk}\}-\min\{ES_{1k},\dots,ES_{nk}\}}$$

```{r calculate ssgsea for PTB vs Latent, cache=TRUE, results = "hide"}
# Select sample contain information about PTB and Latent
# Modify the list to make it compatible with TBSignatureProfiler

multi_set_PTB_Latent <- lapply(object_match, function(x)
  SubsetTBStatus(x, annotationColName = "TBStatus",diseases = c("Latent","PTB"), experiment_type = "assay_reduce")) %>% plyr::compact()   # Remove empty objects use plyr::compact()

ssgsea_PTB_Latent <- lapply(multi_set_PTB_Latent,
                    function(x) TBSignatureProfiler::runTBsigProfiler(input = x,
                                  useAssay = assayNames(x),
                                  signatures = TBsignatures, 
                                  algorithm = "ssGSEA",
                                  combineSigAndAlgorithm = TRUE,
                                  parallel.sz = 14))


ssgsea_PTB_Latent <- readRDS("~/Desktop/ssgsea_PTB_Latent.RDS")
# Obtain p.value, AUC
ssgsea_PTB_Latent_combine <- combine_auc(ssgsea_PTB_Latent, annotationColName = "TBStatus", signatureColNames = names(TBsignatures), num.boot = 50, percent = 0.95)
```

### Boxplot of all scores signatures{.tabset}

```{r Boxplot for each signature PTB vs. Latent, cache=TRUE, results = "asis", fig.width=28, fig.height = 30, fig.wide=TRUE}
for (i in names(TBsignatures)){

  cat("####", i, "\n")
  
  BoxplotTBSig(ssgsea_PTB_Latent, gset = i, annotationColName = "TBStatus")
  
  cat("\n\n")
}
```

```{r include = FALSE}
# Need this chunk to ensure the inclusion/embedding of DT required javascript to work 
DT::datatable(NULL)
```

### Table with T-tests & AUC{.tabset}

```{r Table for each signature PTB vs. Latent, cache=TRUE, results = "asis"}
for (i in names(TBsignatures)){

  cat("####", i, "\n")
  
  print(htmltools::tagList(DT::datatable(ssgsea_PTB_Latent_combine %>% dplyr::filter(Signature == i) %>% dplyr::select(-Signature))))
  
  cat("\n\n")
}
```

### Summary Table for the performance of each signature
Compute mean AUC and bootstrapped 95% CI
```{r suumary table PTB vs. Latent}
# Seletct signatures and associated AUC
PTB_Latent_auc_summary <- ssgsea_PTB_Latent_combine %>% dplyr::select(Signature, AUC) %>% dplyr::group_split(Signature)

# Get summarized table and bootstrap 95% Confidence Interval
cbind(Signature=sapply(PTB_Latent_auc_summary, function(x) x$Signature[1]),
      do.call(rbind,lapply(PTB_Latent_auc_summary, function(x){
  bootstrap_mean_CI(x, colName="AUC", percent=0.95, num.boot=100)
}
)) ) %>% DT::datatable()

```

### Comparison of Signature Performance using AUCs{.tabset}

#### Ridge Plot

```{r AUC ridge plot for PTB vs Latent, fig.width=8, fig.height = 15, cache=TRUE}
# Ridge plot for AUCs distribution across datasets
# PTB vs. Latent
get_auc_distribution(ssgsea_PTB_Latent_combine) + ggtitle("Ridge plot of AUC for PTB vs. Latent") + theme(axis.text.x = ggplot2::element_text(colour="Black", size=12, hjust = 0.5, face="bold"),
                 axis.text.y = ggplot2::element_text(size=12, angle = 0, hjust = 0.5))

```

#### Heatmap{.tabset}

##### Without grouping
```{r PTB vs. Latent heatmap without grouping, fig.width=10, fig.height = 15, fig.wide=TRUE}
# Import signature and data information
GSE_sig <- readxl::read_excel("~/Desktop/RA work/build_TB_data/Signatures/Signature_data_cur.xlsx",sheet = "Sheet2")

heatmap_auc(ssgsea_PTB_Latent_combine, GSE_sig, names(TBsignatures), facet = FALSE) + ggtitle("Heatmap of AUC for PTB vs. Latent")
```

##### With grouping
```{r PTB vs. Latent heatmap with grouping, fig.width=10, fig.height = 15, fig.wide=TRUE}
# Import signature and data information
heatmap_auc(ssgsea_PTB_Latent_combine, GSE_sig, names(TBsignatures), facet = TRUE) + ggtitle("Heatmap of AUC for PTB vs. Latent")
```

## Avtive TB (PTB) vs. Other Diseases (OD) {.tabset}

### Calculate signature scores using ssGSEA
```{r calculate ssgsea for PTB vs OD, cache=TRUE, results = "hide"}
# Select sample contain information about PTB and Other Diseases (OD)
# Modify the list to make it compatible with TBSignatureProfiler

multi_set_PTB_OD <- lapply(object_match, function(x)
  SubsetTBStatus(x, annotationColName = "TBStatus",diseases = c("PTB","OD"), experiment_type = "assay_reduce")) %>% plyr::compact()   # Remove empty objects use plyr::compact()

ssgsea_PTB_OD <- lapply(multi_set_PTB_OD,
                    function(x) TBSignatureProfiler::runTBsigProfiler(input = x,
                                  useAssay = assayNames(x),
                                  signatures = TBsignatures, 
                                  algorithm = "ssGSEA",
                                  combineSigAndAlgorithm = TRUE,
                                  parallel.sz = 14))


# Obtain p.value, AUC
ssgsea_PTB_OD_combine <- combine_auc(ssgsea_PTB_OD, annotationColName = "TBStatus", signatureColNames = names(TBsignatures), num.boot=50)

```

### Boxplots signature scores{.tabset}

```{r Boxplot for each signature PTB vs. OD, cache=TRUE, results = "asis", fig.width=28, fig.height = 30, fig.wide=TRUE}
for (i in names(TBsignatures)){

  cat("####", i, "\n")
  
  BoxplotTBSig(ssgsea_PTB_OD, gset = i, annotationColName = "TBStatus")
  
  cat("\n\n")
}
```

```{r include = FALSE}
# Need this chunk to ensure the inclusion/embedding of DT required javascript to work 
DT::datatable(NULL)
```

### Table with T-tests & AUC{.tabset}

```{r Table for each signature PTB vs. OD, cache=TRUE, results = "asis"}
for (i in names(TBsignatures)){

  cat("####", i, "\n")
  
  print(htmltools::tagList(DT::datatable(ssgsea_PTB_OD_combine %>% dplyr::filter(Signature == i) %>% dplyr::select(-Signature))))
  
  cat("\n\n")
}
```

### Summary Table for the performance of each signature

Evaluate AUCs by taking te means.
```{r suumary table PTB vs. OD}
ssgsea_PTB_OD_combine %>% dplyr::group_by(Signature) %>% dplyr::summarise_all(mean) %>% select(Signature, AUC, LowerAUC, UpperAUC) %>% DT::datatable()
```

### Comparison of Signature Performance using AUCs{.tabset}

#### Ridge Plot

```{r AUC ridge plot for PTB vs OD, fig.width=8, fig.height = 15, cache=TRUE}
# Ridge plot for AUCs distribution across datasets
# PTB vs. OD
get_auc_distribution(ssgsea_PTB_OD_combine) + ggtitle("Ridge plot of AUC for PTB vs. OD") + ggplot2::theme(axis.text.x = ggplot2::element_text(colour="Black", size=12, hjust = 0.5, face="bold"),
                 axis.text.y = ggplot2::element_text(size=12, angle = 0, hjust = 0.5))

```

#### Heatmap{.tabset}

##### Without grouping
```{r PTB vs. OD heatmap without grouping, fig.width=8, fig.height = 12, fig.wide=TRUE}
# Import signature and data information
GSE_sig <- readxl::read_excel("~/Desktop/RA work/build_TB_data/Signatures/Signature_data_cur.xlsx",sheet = "Sheet2")

heatmap_auc(ssgsea_PTB_OD_combine, GSE_sig, names(TBsignatures), facet = FALSE) + ggtitle("Heatmap of AUC for PTB vs. OD")
```

##### With grouping
```{r PTB vs. OD heatmap with grouping, fig.width=8, fig.height = 12, fig.wide=TRUE}
heatmap_auc(ssgsea_PTB_OD_combine, GSE_sig, names(TBsignatures), facet = TRUE) + ggtitle("Heatmap of AUC for PTB vs. OD")

# Use if want to add axis label
# + xlab("Datasets") + theme(axis.title.x = element_text(angle=45,vjust=0.5,hjust=-0.12,size=13,face="bold"))
```



## Signature for predicting progression{.tabset}

Datasets GSE79362, GSE107993, GSE107994, GSE112104 contain information about latent progression. Next, we consider using gene signatures to predict progression information. GSE107993 only includes latent non-progressor and control samples.

### Calculate ssGSEA

```{r Calculate ssGSEA for progression, cache=TRUE, results = "hide"}
gse_progress <- c("GSE79362", "GSE107994", "GSE112104")
multi_set_progress <- lapply(object_match[gse_progress], function(x)
  SubsetTBStatus(x, annotationColName = "Progression", c("Positive","Negative"), experiment_type = "assay_reduce")) %>% plyr::compact()

ssgsea_progress <- lapply(multi_set_progress,
                    function(x) TBSignatureProfiler::runTBsigProfiler(input = x,
                                  useAssay = assayNames(x),
                                  signatures = TBsignatures, 
                                  algorithm = "ssGSEA",
                                  combineSigAndAlgorithm = TRUE,
                                  parallel.sz = 14))

# Obtain p.value, AUC
ssgsea_progress_combine <- combine_auc(ssgsea_progress, annotationColName = "Progression", signatureColNames = names(TBsignatures), num.boot=50)
```

### Boxplot for progression{.tabset}

```{r Boxplot for progression, results = "asis", fig.width=10, fig.height = 6, fig.wide=TRUE}
for (i in names(TBsignatures)){

  cat("####", i, "\n")
  
  BoxplotTBSig(ssgsea_progress, gset = i, annotationColName = "Progression")
  
  cat("\n\n")
}
```

```{r include = FALSE}
# Need this chunk to ensure the inclusion/embedding of DT required javascript to work 
DT::datatable(NULL)
```

### Table for progression{.tabset}

```{r Table for progression, results = "asis"}
for (i in gse_progress){

  cat("####", i, "\n")
  
  print(htmltools::tagList(DT::datatable(ssgsea_progress_combine %>% dplyr::filter(GSE == i) %>% dplyr::select(-GSE))))

  cat("\n\n")
}
```

### Summary table
```{r summary for progression}
ssgsea_progress_combine %>% dplyr::group_by(Signature) %>% dplyr::summarise_all(mean) %>% select(Signature, AUC, LowerAUC, UpperAUC) %>% DT::datatable()
```

### Heatmap{.tabset}

#### Without Grouping

```{r progression heatmap without grouping, fig.width=6, fig.height = 10}
heatmap_auc(ssgsea_progress_combine, GSE_sig, names(TBsignatures), facet = FALSE) + xlab("Datasets") + theme(axis.title.x = element_text(angle=45,vjust=1.3,hjust=-0.2,size=12,face="bold")) + ggtitle("Heatmap of AUC for predicting TB progression")
```

#### With Group

```{r progression heatmap with grouping, fig.width=6, fig.height = 10}
heatmap_auc(ssgsea_progress_combine, GSE_sig, names(TBsignatures), facet = TRUE) + ggtitle("Heatmap of AUC for predicting TB progression") + xlab("Datasets") + theme(axis.title.x = element_text(angle=45,vjust=1.3,hjust=-0.2,size=12,face="bold")) + xlab("Datasets")

```


```{r session info}
sessionInfo()
```
