---
title: "GSE107994"
author: "Xutao Wang"
date: "10/26/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries}
library(GEOquery)
library(preprocessCore)
library(SummarizedExperiment)
library(dplyr)
library(readxl)
# Illumina HiSeq 4000 (Homo sapiens)
```

## READ in data
```{r}
urls <-  getGEOSuppFiles("GSE107994",fetch_files = FALSE)
url_non_normalized <- as.character(urls$url[1])
temp <- tempfile()
download.file(url_non_normalized,temp)

library(readxl)
GSE107994_Non_normalized <- read_excel(temp)
colnames(GSE107994_Non_normalized)
# Obtain row data information from the raw data
row_info <- c('Genes','Gene_name','Gene_biotype')
row_data <- GSE107994_Non_normalized %>% dplyr::select(row_info) %>% DataFrame()
row.names(row_data) <- GSE107994_Non_normalized$Genes
```

```{r}
##  remove row information from the original data and get column data information
gse <- getGEO("GSE107994", GSEMatrix =  F)

GSE107994_raw_counts <- GSE107994_Non_normalized %>% dplyr::select(-row_info) %>% data.frame()
colnames(GSE107994_raw_counts) <- names(GSMList(gse)) 
row.names(GSE107994_raw_counts) <- GSE107994_Non_normalized$Genes

# Create Column Data
data_characteristic <- lapply(1:length(GSMList(gse)), function(x)  
  GSMList(gse)[[x]]@header$characteristics_ch1)
characteristic_table <- sapply(1:length(data_characteristic[[1]]), function(x) 
  sapply(data_characteristic, '[[',x)) 
head(characteristic_table)
## Demo convert geographical region: Kenya to Kenya
characteristic_data_frame <- sub('(.*?): ','',characteristic_table) %>% DataFrame()
#colnames(characteristic_data_frame) <- gsub(':.*','',data_characteristic[[1]])
colnames(characteristic_data_frame) <- c('Tissue','TBStatus','Outlier',
                                         'PatientID','TBType','Smear_result',
                                         'Ethnicity','BirthRegion',"uk_arrival_year",
                                         'Gender','Age','Visit_date','Timepoint_month')

row.names(characteristic_data_frame) <- names(GSMList(gse)) 
head(characteristic_data_frame)

```

## Create Summarized Experiment
```{r}
# Create Metadata
GSE107994_experimentData <- new('MIAME',
                      name="Akul Singhania",
                      lab="The Francis Crick Institute",
                      contact="akul.singhania@crick.ac.uk",
                      title="A modular transcriptional signature identifies phenotypic heterogeneity of human tuberculosis infection.",
                      abstract="We undertook RNA Sequencing (RNA-Seq) of our earlier Berry et al. 2010 (GSE19444 and GSE19442) cohorts and additionally set up a prospective cohort study at Leicester (UK) in subject groups of incident TB and recent TB contacts, respectively. In the Leicester cohort, we performed systematic longitudinal sampling and clinical characterisation first, to validate our TB signature using RNA-Seq in a new and independent cohort of individuals with active TB and LTBI, and secondly to provide longitudinal data in a low TB incidence setting.",
                      pubMedIds = '29921861',
                      other=list(Platform = 'Illumina HiSeq 4000 (Homo sapiens)'))


GSE107994_summarizedExperiemnt <- SummarizedExperiment(assays = list(GSE107994_Non_normalized_counts= as.matrix(GSE107994_raw_counts)), colData = characteristic_data_frame, rowData = row_data, metadata = list(GSE107994_experimentData)); GSE107994_summarizedExperiemnt

```

```{r}
saveRDS(GSE107994_summarizedExperiemnt, ascii=FALSE, file='GSE107994/GSE107994_summarizedExperiemnt.RDS')

# Remove files in temporary directory
unlink(paste0(normalizePath(tempdir()), "/", dir(tempdir())), recursive = TRUE)
dir(tempdir()) # check the temporary directory is empty
```

### Make a matrix contain the NA's for the TB status cariter
#### Standard column names come from Berry GSE19443
```{r}
getwd()
geo_access <- 'GSE107994'
sobject <- readRDS(paste0(geo_access,'/',geo_access,'_summarizedExperiemnt.RDS'))
```

```{r create new column information data}
standard_name_seq <- colData(readRDS('~/Desktop/RA work/build_TB_data/Signatures/Berry/GSE19443/GSE19443_summarizedExperiemnt.RDS')) %>% data.frame() %>% colnames
new_coldata <- function(sobject){
  col_info <- colData(sobject) %>% data.frame()
  dat_NA_new <- matrix(c(rep('n.a.',length(standard_name_seq)*nrow(col_info))),ncol=length(standard_name_seq),
                             byrow=T,dimnames=list(row.names(col_info),
                                                   standard_name_seq)) %>% data.frame()
  
  # fill in with existing data
  overlap_name <- standard_name_seq[which(colnames(dat_NA_new) %in% colnames(col_info))]
  for (i in overlap_name){
    dat_NA_new[i] <- col_info[i]
  }
  
  # Append the rest of cloumns to the dataframe
  col_info_rest <- col_info %>% dplyr::select(-overlap_name)
  dat_final <- cbind(dat_NA_new,col_info_rest)
  
  return(dat_final)
}

new_col_info <- new_coldata(sobject)

```

# Rename TBstatus
### PTB: active TB
### Latent
### Control
### OD: other disease
```{r}
new_col_info$Gender <- ifelse(new_col_info$Gender=="M","Male","Female")
new_col_info$Notes <- new_col_info$TBStatus
TBStatus <- TBStatus_temp <- as.character(new_col_info$TBStatus)
unique(TBStatus_temp)

TBStatus[grep("Active_TB",TBStatus_temp)] <- "PTB"
TBStatus[which(TBStatus %in% c("LTBI_Progressor","LTBI"))] <- "Latent"

TBStatus

new_col_info$TBStatus <- TBStatus

new_col_info %>% View()

colData(sobject) <- new_col_info %>% DataFrame()

saveRDS(sobject,paste0(geo_access,'/',geo_access,'_sobject.RDS'))

```



