---
title: "GSE62525"
author: "Xutao Wang"
date: "10/28/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries}
library(GEOquery)
library(preprocessCore)
library(SummarizedExperiment)
library(dplyr)
library(limma)

# Phalanx Human OneArray Ver. 6 Release 1
```

```{r}
urls <-  getGEOSuppFiles("GSE62525",fetch_files = FALSE)
url_cel <- as.character(urls$url[1])
temp <- tempfile()
tempd <- tempdir()
download.file(url_cel,temp)
untar(temp,exdir = tempd)

gprFiles <- list.files(path = tempd, pattern = '*.gpr',full.names=TRUE);gprFiles
GSE62525_RG <- read.maimages(gprFiles, source="genepix",columns = list(R = "F635 Median", G = "B635 Median"))
```

## Create Row Data
```{r}
## Create Row Data
row_data <- GSE62525_RG$genes %>% DataFrame()
row.names(row_data) <- paste0(row_data$Block,'_',row_data$Row,'_',row_data$Column)
head(row_data)
```

## Create Non-normalized counts
```{r}
gse <- getGEO("GSE62525", GSEMatrix = F)

GSE62525_raw <- GSE62525_RG$R-GSE62525_RG$G # Forefround - Background
GSE62525_Non_normalized_counts <- GSE62525_raw

colnames(GSE62525_Non_normalized_counts) <- names(GSMList(gse))
row.names(GSE62525_Non_normalized_counts) <- row.names(row_data)
```


## Create Column Data
```{r}
# Create Column Data
data_characteristic <- lapply(1:length(GSMList(gse)), function(x)  
  GSMList(gse)[[x]]@header$characteristics_ch1)
Title <- data_characteristic <- sapply(1:length(GSMList(gse)), function(x)  
  GSMList(gse)[[x]]@header$title)
characteristic_table <- sapply(1:length(data_characteristic[[1]]), function(x) 
  sapply(data_characteristic, '[[',x)) 
head(characteristic_table)
## Demo convert geographical region: Kenya to Kenya
characteristic_data_frame <- sub('(.*?): ','',characteristic_table) %>% as_tibble() %>% mutate(Title=Title) %>% DataFrame()
#colnames(characteristic_data_frame) <- gsub(':.*','',data_characteristic[[1]])
colnames(characteristic_data_frame) <- c('TBStatus','Title')
row.names(characteristic_data_frame) <- names(GSMList(gse))
```

```{r}
# Create Metadata
GSE62525_experimentData <- new('MIAME',
                      name="Julia Tzu-Ya Weng",
                      lab="Yuan Ze University",
                      contact="julweng@gmail.com",
                      title="Gene expression profiling identifies candidate biomarkers for active and latent tuberculosis. ",
                      abstract="We isolated total RNA from the PBMC from 7 active TB, 7 latent infection, and 7 healthy individuals and profiled their transcriptional profiles to identify signficantly differentially expressed geens that differ among these three groups",
                      url="10.1186/s12859-015-0848-x",
                      pubMedIds = '26818387',
                      other=list(Platform = 'Phalanx Human OneArray Ver. 6 Release 1 :GPL1695'))


GSE62525_summarizedExperiemnt <- SummarizedExperiment(assays = list(GSE62525_Non_normalized_counts= as.matrix(GSE62525_Non_normalized_counts)), colData = characteristic_data_frame, rowData = row_data, metadata = list(GSE62525_experimentData)); GSE62525_summarizedExperiemnt

```

```{r}
saveRDS(GSE62525_summarizedExperiemnt, ascii=FALSE, file='GSE62525/GSE62525_summarizedExperiemnt.RDS')

# Remove files in temporary directory
unlink(paste0(normalizePath(tempdir()), "/", dir(tempdir())), recursive = TRUE)
dir(tempdir()) # check the temporary directory is empty
```

### Make a matrix contain the NA's for the TB status cariter
#### Standard column names come from Berry GSE19443
```{r}
getwd()
geo_access <- 'GSE62525'
sobject <- readRDS(paste0(geo_access,'/',geo_access,'_summarizedExperiemnt.RDS'))
```

```{r create new column information data}
standard_name_seq <- colData(readRDS('~/Desktop/RA work/build_TB_data/Signatures/Berry/GSE19443/GSE19443_summarizedExperiemnt.RDS')) %>% data.frame() %>% colnames
new_coldata <- function(sobject){
  col_info <- colData(sobject) %>% data.frame()
  dat_NA_new <- matrix(c(rep('n.a.',length(standard_name_seq)*nrow(col_info))),ncol=length(standard_name_seq),
                             byrow=T,dimnames=list(row.names(col_info),
                                                   standard_name_seq)) %>% data.frame()
  
  # fill in with existing data
  overlap_name <- standard_name_seq[which(colnames(dat_NA_new) %in% colnames(col_info))]
  for (i in overlap_name){
    dat_NA_new[i] <- col_info[i]
  }
  
  # Append the rest of cloumns to the dataframe
  col_info_rest <- col_info %>% dplyr::select(-overlap_name)
  dat_final <- cbind(dat_NA_new,col_info_rest)
  
  return(dat_final)
}

new_col_info <- new_coldata(sobject)

```

# Rename TBstatus
### PTB: active TB
### Latent
### Control
### OD: other disease
```{r}
new_col_info$Notes <- new_col_info$Title
new_col_info <- new_col_info[,-which(colnames(new_col_info) %in% 'Title')]
new_col_info$Tissue <- "PBMC"
# Demo:  "PBMC_healthy_C1_rep1" to C1_rep1
new_col_info$Batch <- gsub("^(?:[^_]+_){2}", "\\1",new_col_info$TBStatus)

TBStatus <- TBStatus_temp <- as.character(new_col_info$TBStatus)
unique(TBStatus_temp)

TBStatus[grep("healthy",TBStatus_temp)] <- "Control"
TBStatus[grep("latent",TBStatus_temp)] <- "Latent"
TBStatus[grep("active TB",TBStatus_temp)] <- "PTB"

TBStatus

new_col_info$TBStatus <- TBStatus

new_col_info %>% View()

colData(sobject) <- new_col_info %>% DataFrame()

```

# Edit Row data
```{r}
row_data <- rowData(sobject) %>% data.frame()
View(row_data)

row_data$SYMBOL_NEW <- row_data$Name
row_data$SYMBOL_NEW <- ifelse(row_data$SYMBOL_NEW %in% NA, "NA", row_data$SYMBOL_NEW)

View(row_data)
rowData(sobject) <- row_data %>% DataFrame()

saveRDS(sobject,paste0(geo_access,'/',geo_access,'_sobject.RDS'))
```





















































































