---
title: "GSE36238"
author: "Xutao Wang"
date: "10/1/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, message = FALSE, cache=TRUE}
library(GEOquery)
library(preprocessCore)
library(SummarizedExperiment)
library(affy)
library(dplyr)
library(hgu133plus2.db) # Affymetrix Human Genome U133 Plus 2.0 Array
```

## Read in Non-normalized data
```{r, cache = TRUE}
#setwd('~/Desktop/RA work/build_TB_data/Signatures/Blankley')
#getwd()
urls <-  getGEOSuppFiles("GSE36238",fetch_files = FALSE)
url_cel <- as.character(urls$url[1])
temp <- tempfile()
tempd <- tempdir()
download.file(url_cel,temp)
untar(temp,exdir = tempd)

#untar('GSE36238/GSE36238_RAW.tar', exdir="GSE36238/GSE36238_Raw")
celFiles <- list.files(path = tempd, pattern = '*.CEL',full.names=TRUE);celFiles
GSE36238_data.affy <- ReadAffy(filenames = celFiles)

# Conduct RMA
#GSE54992_Non_normalized <- GSE54992_data.affy@assayData$exprs
GSE36238_normalized_rma <- exprs(affy::rma(GSE36238_data.affy)) # A matrix

 # Convert GSM1327554.CEL.gz to GSM1327554
colnames(GSE36238_normalized_rma) <- sapply(1:ncol(GSE36238_normalized_rma), 
                                          function(x) gsub('\\..*','',colnames(GSE36238_normalized_rma)[x]))
GSE36238_normalized_counts <- GSE36238_normalized_rma
GSE36238_normalized_counts[1:5,1:5]
```

```{r convert ProbeID to Gene Symbol, cache=TURE}
# Read in data from vendor's information
GPL570 <- getGEO('GPL570', GSEMatrix = F)
GPL570_annotate <- GPL570@dataTable@table

# Annotation
PROBES <- row.names(GSE36238_normalized_rma) 
OUT <- AnnotationDbi::select(hgu133plus2.db, PROBES, "SYMBOL")
OUT[is.na(OUT)] <- NA

# Map ProbeID to Gene Symbol 
OUT_collapse <- OUT %>% group_by(PROBEID) %>%  # Multiple gene symbols map to the same ProbeID
  summarise(SYMBOL = paste(SYMBOL, collapse="///"), times = length(unlist(strsplit(SYMBOL, "///")))) 

GSE36238_normalized_rma <- GSE36238_normalized_rma %>% as_tibble() %>% mutate(ID_REF=row.names(GSE36238_normalized_rma))

GSE36238_final <- GSE36238_normalized_rma %>% left_join(OUT_collapse, by=c('ID_REF'="PROBEID")) %>% left_join(GPL570_annotate, by = c('ID_REF'='ID'))

row_data <- GSE36238_final %>% dplyr::select(-grep('GSM',colnames(GSE36238_final))) %>% DataFrame()
# arrange(GSE36238_final, desc(times))
```

```{r create column data}
gse <- getGEO('GSE36238', GSEMatrix =  F)
data_characteristic <- lapply(1:length(GSMList(gse)), function(x)  
  GSMList(gse)[[x]]@header$characteristics_ch1)
characteristic_table <- sapply(1:length(data_characteristic[[1]]), function(x) 
  sapply(data_characteristic, '[[',x)) 
## Demo convert geographical region: Kenya to Kenya
characteristic_data_frame <- sub('(.*?): ','',characteristic_table) %>% DataFrame()
#colnames(characteristic_data_frame) <- gsub(':.*','',data_characteristic[[1]])
# Get string between two words
#characteristic_data_frame$TimePostTreatment <- gsub('.*for (.+).*','\\1',characteristic_data_frame$V1)
#characteristic_data_frame$TimePostTreatment[-grep('month',characteristic_data_frame$TimePostTreatment)] <-'n.a.'

#characteristic_data_frame$TBStatus <- plyr::revalue(characteristic_data_frame$V1,c('tuberculosis'='TB','TB patient after anti-tuberculosis treatment for 3 months'='TB_treatment','TB patient after anti-tuberculosis treatment for 6 months'='TB_treatment','latent tuberculosis infection'='latent','healthy donor'='control'))
#characteristic_data_frame$Tissue <- characteristic_data_frame$V2
#characteristic_data_frame <- characteristic_data_frame[,-c(1,2)]
colnames(characteristic_data_frame) <- c('TimePoint','Tissue','Pateient_id','SampleTime')
row.names(characteristic_data_frame) <- names(GSMList(gse))
head(characteristic_data_frame)
```

```{r Create RowData MetaData and summarized experiment}
# Create Metadata
GSE36238_experimentData <- new('MIAME',
                      name="Jackie Cliff",
                      lab="London School of Hygiene & Tropical Medicine",
                      contact="jackie.cliff@lshtm.ac.uk",
                      title="Distinct phases of blood gene expression pattern through tuberculosis treatment reflect modulation of the humoral immune response.",
                      abstract="An example SummarizedExperimentObject",
                      url="10.1093/infdis/jis499",
                      pubMedIds = '22872737',
                      other=list(Platform = 'Affymetrix Human Genome U133 Plus 2.0 Array: GPL570'))

GSE36238_summarizedExperiemnt <- SummarizedExperiment(assays = list(GSE36238_normalized_counts= as.matrix(GSE36238_normalized_counts)), colData = characteristic_data_frame, rowData = row_data, metadata = list(GSE36238_experimentData)); GSE36238_summarizedExperiemnt

```

```{r}
saveRDS(GSE36238_summarizedExperiemnt, ascii=FALSE, file='GSE36238/GSE36238_summarizedExperiemnt.RDS')

# Remove files in temporary directory
unlink(paste0(normalizePath(tempdir()), "/", dir(tempdir())), recursive = TRUE)
dir(tempdir()) # check the temporary directory is empty
```

### Make a matrix contain the NA's for the TB status cariter
#### Standard column names come from Berry GSE19443
```{r}
getwd()
geo_access <- 'GSE36238'
sobject <- readRDS(paste0(geo_access,'/',geo_access,'_summarizedExperiemnt.RDS'))
```

```{r create new column information data}
standard_name_seq <- colData(readRDS('~/Desktop/RA work/build_TB_data/Signatures/Berry/GSE19443/GSE19443_summarizedExperiemnt.RDS')) %>% data.frame() %>% colnames
new_coldata <- function(sobject){
  col_info <- colData(sobject) %>% data.frame()
  dat_NA_new <- matrix(c(rep('n.a.',length(standard_name_seq)*nrow(col_info))),ncol=length(standard_name_seq),
                             byrow=T,dimnames=list(row.names(col_info),
                                                   standard_name_seq)) %>% data.frame()
  
  # fill in with existing data
  overlap_name <- standard_name_seq[which(colnames(dat_NA_new) %in% colnames(col_info))]
  for (i in overlap_name){
    dat_NA_new[i] <- col_info[i]
  }
  
  # Append the rest of cloumns to the dataframe
  col_info_rest <- col_info %>% dplyr::select(-overlap_name)
  dat_final <- cbind(dat_NA_new,col_info_rest)
  
  return(dat_final)
}
new_col_info <- new_coldata(sobject)
```

# Rename TBstatus
### PTB: active TB
### Latent
### Control
### OD: other disease
```{r}
new_col_info$TBStatus <- 'PTB'
new_col_info %>% View()

colData(sobject) <- new_col_info %>% DataFrame()
saveRDS(sobject,paste0(geo_access,'/',geo_access,'_sobject.RDS'))

```
