---
title: "GSE31348"
author: "Xutao Wang"
date: "10/1/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, message = FALSE, cache=TRUE}
library(GEOquery)
library(preprocessCore)
library(SummarizedExperiment)
library(affy)
library(dplyr)
library(hgu133plus2.db) # Affymetrix Human Genome U133 Plus 2.0 Array
```

## Read in Non-normalized data
```{r, cache = TRUE}
setwd('~/Desktop/RA work/build_TB_data/Signatures/Blankley')
getwd()
geo_access <- "GSE31348"
#filePaths = getGEOSuppFiles(geo_access)
#untar(paste0(geo_access,'/',geo_access,'_RAW.tar'), exdir=paste0(geo_access,'/',geo_access,'_Raw'))
## Cannot untar the SuppFiles, manually Download from GEO
celFiles <- list.celfiles(path = paste0(geo_access,'/',geo_access,'_Raw'), full.names=TRUE)
GSE31348_data.affy <- ReadAffy(filenames = celFiles)

# Conduct RMA
#GSE54992_Non_normalized <- GSE54992_data.affy@assayData$exprs
GSE31348_normalized_rma <- exprs(rma(GSE31348_data.affy)) # A matrix

# Convert GSM1327554.CEL.gz to GSM1327554
colnames(GSE31348_normalized_rma) <- sapply(1:ncol(GSE31348_normalized_rma), 
                                          function(x) gsub('_.*','',colnames(GSE31348_normalized_rma)[x]))
GSE31348_normalized_counts <- GSE31348_normalized_rma
GSE31348_normalized_counts[1:5,1:5]
```

```{r convert ProbeID to Gene Symbol, cache=TURE}
# Annotation
PROBES <- row.names(GSE31348_normalized_rma) 
OUT <- AnnotationDbi::select(hgu133plus2.db, PROBES, "SYMBOL")
OUT[is.na(OUT)] <- 'Unmapped'

# Map ProbeID to Gene Symbol 
OUT_collapse <- OUT %>% group_by(PROBEID) %>%  # Multiple gene symbols map to the same ProbeID
  summarise(SYMBOL = paste(SYMBOL, collapse=", "), times = length(unlist(strsplit(SYMBOL, ",")))) 

GSE31348_normalized_rma <- GSE31348_normalized_rma %>% as_tibble() %>% mutate(ID_REF=row.names(GSE31348_normalized_rma))
GSE31348_final <- OUT_collapse %>% left_join(GSE31348_normalized_rma, by=c("PROBEID" = 'ID_REF')) %>% as_tibble()

arrange(GSE31348_final, desc(times))
```

```{r create column data}
gse <- getGEO(geo_access, GSEMatrix =  F)
data_characteristic <- lapply(1:length(GSMList(gse)), function(x)  
  GSMList(gse)[[x]]@header$characteristics_ch1)
characteristic_table <- sapply(1:length(data_characteristic[[1]]), function(x) 
  sapply(data_characteristic, '[[',x)) 
## Demo convert geographical region: Kenya to Kenya
characteristic_data_frame <- sub('(.*?): ','',characteristic_table) %>% DataFrame()
#colnames(characteristic_data_frame) <- gsub(':.*','',data_characteristic[[1]])
# Get string between two words
#characteristic_data_frame$TimePostTreatment <- gsub('.*for (.+).*','\\1',characteristic_data_frame$V1)
#characteristic_data_frame$TimePostTreatment[-grep('month',characteristic_data_frame$TimePostTreatment)] <-'n.a.'

#characteristic_data_frame$TBStatus <- plyr::revalue(characteristic_data_frame$V1,c('tuberculosis'='TB','TB patient after anti-tuberculosis treatment for 3 months'='TB_treatment','TB patient after anti-tuberculosis treatment for 6 months'='TB_treatment','latent tuberculosis infection'='latent','healthy donor'='control'))
#characteristic_data_frame$Tissue <- characteristic_data_frame$V2
#characteristic_data_frame <- characteristic_data_frame[,-c(1,2)]
colnames(characteristic_data_frame) <- c('TBStatus','Patient_id','TimePoint','Tissue')
row.names(characteristic_data_frame) <- colnames(GSE31348_normalized_counts)
head(characteristic_data_frame)
```

```{r Create RowData MetaData and summarized experiment}
# Create Row data annotation
row_data <- DataFrame(SYMBOL=GSE31348_final$SYMBOL)

# Create Metadata
GSE31348_experimentData <- new('MIAME',
                      name="Jackie Cliff",
                      lab="London School of Hygiene & Tropical Medicine",
                      contact="jackie.cliff@lshtm.ac.uk",
                      title="Distinct phases of blood gene expression pattern through tuberculosis treatment reflect modulation of the humoral immune response.",
                      abstract="An example SummarizedExperimentObject",
                      url="10.1093/infdis/jis499",
                      pubMedIds = '22872737',
                      other=list(Platform = 'Affymetrix Human Genome U133 Plus 2.0 Array'))

GSE31348_summarizedExperiemnt <- SummarizedExperiment(assays = list(GSE31348_normalized_counts= as.matrix(GSE31348_normalized_counts)), colData = characteristic_data_frame, rowData = row_data, metadata = list(GSE31348_experimentData)); GSE31348_summarizedExperiemnt

```

```{r}
saveRDS(GSE31348_summarizedExperiemnt, ascii=FALSE, file='GSE31348/GSE31348_summarizedExperiemnt.RDS')
```
